Sources/NIO/ConvenienceOptionSupport.swift

```diff
@@ -0,0 +1,178 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2020 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// MARK:  Universal Client Bootstrap
+extension NIOClientTCPBootstrapProtocol {
+    /// Apply any understood shorthand options to the bootstrap, removing them from the set of options if they are consumed.
+    /// - parameters:
+    ///     - options:  The options to try applying - the options applied should be consumed from here.
+    /// - returns: The updated bootstrap with and options applied.
+    public func _applyOptions(_ options: inout NIOTCPShorthandOptions) -> Self {
+        // Default is to consume no options and not update self.
+        return self
+    }
+}
+
+extension NIOClientTCPBootstrap {
+    /// Specifies some `ChannelOption`s to be applied to the channel.
+    /// - See: channelOption
+    /// - Parameter options: Set of shorthand options to apply.
+    /// - Returns: The updated bootstrap (`self` being mutated)
+    public func options(_ options: NIOTCPShorthandOptions) -> NIOClientTCPBootstrap {
+        var optionsRemaining = options
+        // First give the underlying a chance to consume options.
+        let withUnderlyingOverrides =
+            NIOClientTCPBootstrap(self, withUpdated: underlyingBootstrap._applyOptions(&optionsRemaining))
+        // Default apply any remaining options.
+        return optionsRemaining.applyFallbackMapping(withUnderlyingOverrides)
+    }
+}
+
+// MARK: Utility
+/// Has an option been set?
+/// Option has a value of generic type T.
+public enum NIOOptionValue<T> {
```

Comment:
@Davidde94 / @glbrntt / @Lukasa shouldn't we nest this under `ChannelOptions.Types`?

Reply:
Makes sense

Reply:
Yes. Down with top-level types that users should never actually use.